### PR TITLE
Fix empty array index lookup

### PR DESCRIPTION
**Pull request by chaudum**
**Description**: What this PR does

This PR fixes a panic that occurs when the array index is an empty string, as in

```
{"foo": {"keys": ["a", "b", "c"]}}
// lookup foo.keys[""] instead of foo.keys[1]
```

### CODE REVIEW

File: parser_test.go [hunk 1]
```diff
@@ -1498,10 +1498,10 @@ func TestArrayEachEmpty(t *testing.T) {
 	}{
 		{"Empty array", args{[]byte("[]"), funcError, []string{}}, 1, false},
 		{"Empty array with space", args{[]byte("[ ]"), funcError, []string{}}, 2, false},
-		{"Empty array with \n", args{[]byte("[\n]"), funcError, []string{}}, 2, false},
+		{"Empty array with newline", args{[]byte("[\n]"), funcError, []string{}}, 2, false},
 		{"Empty field array", args{[]byte("{\"data\": []}"), funcError, []string{"data"}}, 10, false},
 		{"Empty field array with space", args{[]byte("{\"data\": [ ]}"), funcError, []string{"data"}}, 11, false},
-		{"Empty field array with \n", args{[]byte("{\"data\": [\n]}"), funcError, []string{"data"}}, 11, false},
+		{"Empty field array with newline", args{[]byte("{\"data\": [\n]}"), funcError, []string{"data"}}, 11, false},
```
Comment (trevorwhitney) [inline parser_test.go:1504]: do we need a test for array with an empty string, ie `[]byte("{\"data\": [\"\"]}"

File: parser_test.go [hunk 2]
```diff
@@ -1708,89 +1708,80 @@ var testJson = []byte(`{
...
+			path:     []string{"a\n", "b\n"}, // issue#165
+			expected: "99",
+		}, {
+			path:     []string{"nested", "b"}, // Should find repeated key
+			expected: "2",
+		}, {
+			path:     []string{"arrString", "[1]"},
+			expected: "b",
+		}, {
+			path:     []string{"arrString", "[]"}, // [] is equal to [0]
```
Comment (trevorwhitney) [inline parser_test.go:1759]: again, do we need `[]string{"arrString", "[\"\"]"}`, or maybe that case works?

**Review by trevorwhitney — APPROVED** [review]
looks good, just a couple questions about an array with empty string, ie `[""]`, thanks!

File: parser_test.go (hunk 1 above)
Comment (chaudum) [reply parser_test.go:1504]: As I understood, this is just a test to iterate over an array, not accessing an element by index in that array. The API of this library is rather confusing tbh.

File: parser_test.go (hunk 2 above)
Comment (chaudum) [reply parser_test.go:1759]:
```
[]string{"keys", "[]"}
```
is what Loki produces when using this parser expression:

```
| json key[""]
```

---

```
[]string{"keys", "[\"\"]"}
```
would be equivalent to any other invalid, non-integer array index
